### PR TITLE
[2.0.alpha] fixed backend_opts to be empty dictionary instead of None

### LINKED ISSUE #19811: [2.0.alpha] Extensions subgraph example is failing
## Description
Extensions subgraph example is failing for the v2.0 alpha tag with the following error message on Linux (CPU build):
```
-------------------------------
Testing myProp Gluon Hybridize partitioning with extra input
Traceback (most recent call last):
  File "test_subgraph.py", line 105, in <module>
    test("myProp")
  File "test_subgraph.py", line 77, in test
    sym_block2.optimize_for(mx.nd.ones((3,2)), mx.nd.ones((3,2)), backend="addInputPass")
  File "/home/ubuntu/incubator-mxnet/python/mxnet/gluon/block.py", line 1390, in optimize_for
    self._build_cache(x, *args)
  File "/home/ubuntu/incubator-mxnet/python/mxnet/gluon/block.py", line 1203, in _build_cache
    out = out.optimize_for(self._backend, arg_dict, aux_dict, ctx, input_shapes, **self._backend_opts)
TypeError: optimize_for() argument after ** must be a mapping, not NoneType
```
The problem seems to be coming from here:
https://github.com/apache/incubator-mxnet/blob/49edbfc82cb0ce3b823051e1acc26388bc79686c/python/mxnet/gluon/block.py#L1398-L1401
Where we set `_backend_opts` to `None` and later the `**` operator fails since its not a dictionary. This change seems to have been introduced in #19543:
https://github.com/apache/incubator-mxnet/commit/dbcd6a19f1462d5fe78c2174372e2db42409d7b4#diff-62a7f3cb1a7d34ef73f8c048333c5b128cf1ac1461e6729080bebc7505cfe3e4R1264

Changing this to set `_backend_opts` to an empty dictionary seems to fix the problem.

### PR DESCRIPTION
**Pull request by samskalicky**
## Description ##
Fixes #19811

### CODE REVIEW

**Review by leezu — COMMENTED** [review]
Thank you! Could you please add a test? It looks like this was completely untested?

File: python/mxnet/gluon/block.py [hunk 1]
```diff
@@ -1052,7 +1052,7 @@ def __init__(self):
         self._callback = None
         self._monitor_all = False
         self._backend = None
-        self._backend_opts = {}
+        self._backend_opts = dict()
```
Comment (szha) [inline python/mxnet/gluon/block.py:1055]: I always had the impression that initialization with literal empty dict is preferred.

Comment (leezu) [reply python/mxnet/gluon/block.py:1055]: The only exception may be function / method arguments

Comment (samskalicky) [reply python/mxnet/gluon/block.py:1055]: will change to literal

**Comment by samskalicky** [comment]
> Thank you! Could you please add a test? It looks like this was completely untested?

Looking at the flow, it seems to only happen when calling `optimize_for` twice. After each call the `backend` and `backend_opts` are cleared:
https://github.com/apache/incubator-mxnet/blob/49edbfc82cb0ce3b823051e1acc26388bc79686c/python/mxnet/gluon/block.py#L1398-L1401

And then the "clearing" sets the `backend_opts` to None. So on the 2nd call, if the user does not provide any options it would run into this scenario. In the lib_subgraph extensions example we call optimize_for once to partition, and then a 2nd time to run a graph pass:
https://github.com/apache/incubator-mxnet/blob/49edbfc82cb0ce3b823051e1acc26388bc79686c/example/extensions/lib_subgraph/test_subgraph.py#L65-L78

But in the unittests, we only use the `default` subgraph backends:
https://github.com/apache/incubator-mxnet/blob/49edbfc82cb0ce3b823051e1acc26388bc79686c/tests/python/unittest/test_subgraph_op.py#L267

So theres no way to duplicate exactly what is being done in the example. But, we do build the `lib_subgraph` library in the CMakeLists.txt:
https://github.com/apache/incubator-mxnet/blob/49edbfc82cb0ce3b823051e1acc26388bc79686c/CMakeLists.txt#L814

So we could add a test in unittests to use this library. But I think we decided not to do this in the past since people were running the unittests on pip wheels that didnt have the libraries packaged into the wheel. Thoughts?

**Comment by leezu** [comment]
> But I think we decided not to do this in the past since people were running the unittests on pip wheels that didnt have the libraries packaged into the wheel. Thoughts?

The CI does not use wheels for most pipelines AFAIK. You can just add a separate test folder and execute it manually in one of the CI pipelines?

**Comment by leezu** [comment]
@mxnet-bot run ci [unix-cpu, windows-cpu, windows-gpu]

**Comment by samskalicky** [comment]
> The CI does not use wheels for most pipelines AFAIK. You can just add a separate test folder and execute it manually in one of the CI pipelines?

I found a file `test_extensions.py` in tests/python/unittest and added the relevant code from the example so that it will be tested in the CI. Thanks for the discussion!

**Pull request merged**